include/albatross/src/cereal/suitesparse.hpp

```diff
@@ -0,0 +1,486 @@
+// helpers have difficult state properties.
+inline void suitesparse_cereal_realloc(void **p, std::size_t count,
+                                       std::size_t elem_size) {
+  if (nullptr != *p) {
+    free(*p);
+  }
+  *p = malloc(count * elem_size);
+  ALBATROSS_ASSERT(nullptr != *p);
+}
+
+}  // namespace detail
+
+template <class Archive>
+inline void save(Archive &ar, cholmod_common const &cc,
+                 std::uint32_t version ALBATROSS_UNUSED) {
+  ar(CEREAL_NVP(cc.dbound));
+  ar(CEREAL_NVP(cc.grow0));
+  ar(CEREAL_NVP(cc.grow1));
+  ar(CEREAL_NVP(cc.grow2));
+  ar(CEREAL_NVP(cc.maxrank));
+  ar(CEREAL_NVP(cc.supernodal_switch));
+  ar(CEREAL_NVP(cc.supernodal));
+  ar(CEREAL_NVP(cc.final_asis));
+  ar(CEREAL_NVP(cc.final_super));
+  ar(CEREAL_NVP(cc.final_ll));
+  ar(CEREAL_NVP(cc.final_pack));
+  ar(CEREAL_NVP(cc.final_monotonic));
+  ar(CEREAL_NVP(cc.final_resymbol));
+  static_assert(sizeof(cc.zrelax) == 3 * sizeof(double),
+                "cholmod_common::zrelax expected to have 3 elements.");
+  ar(::cereal::make_nvp("cc.zrelax0", cc.zrelax[0]));
+  ar(::cereal::make_nvp("cc.zrelax1", cc.zrelax[1]));
+  ar(::cereal::make_nvp("cc.zrelax2", cc.zrelax[2]));
+  static_assert(sizeof(cc.nrelax) == 3 * sizeof(std::size_t),
+                "cholmod_common::nrelax expected to have 3 elements.");
+  ar(::cereal::make_nvp("cc.nrelax0", cc.nrelax[0]));
+  ar(::cereal::make_nvp("cc.nrelax1", cc.nrelax[1]));
+  ar(::cereal::make_nvp("cc.nrelax2", cc.nrelax[2]));
+  ar(CEREAL_NVP(cc.prefer_zomplex));
+  ar(CEREAL_NVP(cc.prefer_upper));
+  ar(CEREAL_NVP(cc.quick_return_if_not_posdef));
+  ar(CEREAL_NVP(cc.prefer_binary));
+  ar(CEREAL_NVP(cc.print));
+  ar(CEREAL_NVP(cc.precise));
+  ar(CEREAL_NVP(cc.try_catch));
+  // Do not serialise the user error handler.
+  ar(CEREAL_NVP(cc.nmethods));
+  ar(CEREAL_NVP(cc.current));
+  ar(CEREAL_NVP(cc.selected));
+  constexpr std::size_t max_methods = sizeof(cc.method) / sizeof(cc.method[0]);
+  for (std::size_t m = 0; m < max_methods; ++m) {
+    ar(cc.method[m]);
+  }
+  // method
+  ar(CEREAL_NVP(cc.postorder));
+  ar(CEREAL_NVP(cc.default_nesdis));
+  ar(CEREAL_NVP(cc.metis_memory));
+  ar(CEREAL_NVP(cc.metis_dswitch));
+  ar(CEREAL_NVP(cc.metis_nswitch));
+  ar(::cereal::make_nvp("nrow", cc.nrow));
+  ar(CEREAL_NVP(cc.mark));
+  ar(::cereal::make_nvp("iworksize", cc.iworksize));
+  // The comments in `cholmod_core.h` define this as a size in bytes.
+  // The code, however, considers it to be a number of elements.
+  ar(::cereal::make_nvp("xworksize", cc.xworksize));
+
+  const std::size_t cholmod_int_size =
+      cc.itype == CHOLMOD_LONG ? sizeof(SuiteSparse_long) : sizeof(int);
+
+  detail::encode_array(ar, "cc.Flag", cc.Flag, cc.nrow * cholmod_int_size);
+  detail::encode_array(ar, "cc.Head", cc.Head,
+                       (cc.nrow + 1) * cholmod_int_size);
+  detail::encode_array(ar, "cc.Xwork", cc.Xwork, cc.xworksize * sizeof(double));
+
+  // Iwork omitted; must reinitialize
```

Comment:
Not sure I follow this comment?

Reply:
This is related to the previous thing -- I don't serialise the workspaces, I just reallocate new ones when I deserialise.  (I confirmed that this is safe both with tests and by reading the SPQR / cholmod code.)  I'll expand the comment to make it clearer.